src/ircd_parser.y

```diff
@@ -543,6 +545,73 @@ serverinfo_ssl_certificate_file: SSL_CERTIFICATE_FILE '=' QSTRING ';'
       yyerror("RSA private key does not match the SSL certificate public key!");
       break;
     }
+
+    if (ServerInfo.dh_params != NULL &&
+      !SSL_CTX_set_tmp_dh(ServerInfo.ctx, ServerInfo.dh_params))
+    {
+      yyerror(ERR_lib_error_string(ERR_get_error()));
+    }
+  }
+#endif
+};
+
+serverinfo_dh_params_file: DH_PARAMS_FILE '=' QSTRING ';'
+{
+#ifdef HAVE_LIBCRYPTO
+  if (ypass == 1)
+  {
+    BIO *file;
+    int codes;
+
+    if (ServerInfo.dh_params != NULL)
+    {
+      DH_free(ServerInfo.dh_params);
+      ServerInfo.dh_params = NULL;
+    }
+
+    if (ServerInfo.dh_params_file != NULL)
+    {
+      MyFree(ServerInfo.dh_params_file);
+      ServerInfo.dh_params_file = NULL;
+    }
+
+    DupString(ServerInfo.dh_params_file, yylval.string);
+
+    if ((file = BIO_new_file(yylval.string, "r")) == NULL)
+    {
+      yyerror("File open failed, ignoring");
+      break;
+    }
+
+    ServerInfo.dh_params = PEM_read_bio_DHparams(file, NULL,
+      NULL, NULL);
+
+    (void)BIO_set_close(file, BIO_CLOSE);
+    BIO_free(file);
+
+    if (ServerInfo.dh_params == NULL)
+    {
+      yyerror("Couldn't extract DH params, ignoring");
+      break;
+    }
+
+    if (!DH_check(ServerInfo.dh_params, &codes))
```

Comment:
Just a friendly reminder:

DH_check() returning 1 does not mean "all good". The man page says: `DH_check() returns 1 if the check could be performed, 0 otherwise.`

The OpenSSL wiki has a good example of how to use DH_check() correctly, including a small tidbit on how to make it work correctly with IETF groups, like the ones in RFC 2409: https://wiki.openssl.org/index.php/Diffie-Hellman_parameters#Validating_Parameters